extensions/sql/data/JSONSQLWriter.cpp

```diff
@@ -21,78 +21,132 @@
-        valueVal.SetInt64(i64val);
-      }
-      break;
-      case soci::data_type::dt_unsigned_long_long: {
-        uint64_t u64val = row.get<unsigned long long>(i);
-        valueVal.SetUint64(u64val);
-      }
-      break;
-      case soci::data_type::dt_date: {
-        std::tm when = row.get<std::tm>(i);
-        std::string str = std::string(asctime(&when));
-        // asctime adds '\n' in the end.
-        if (str[str.size() - 1] == '\n') {
-          str.resize(str.size() - 1);
+
+    if (row.get_indicator(i) == soci::i_null) {
+      std::string null = "NULL";
+      valueVal.SetString(null.c_str(), null.length(), alloc);
+    } else {
+      switch (const auto dataType = props.get_data_type()) {
+        case soci::data_type::dt_string: {
+          auto value = std::string(row.get<std::string>(i));
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, '\'' + value + '\'');
+          }
+          valueVal.SetString(value.c_str(), value.length(), alloc);
+        }
+        break;
+        case soci::data_type::dt_double: {
+          auto value = row.get<double>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetDouble(value);
+        }
+        break;
+        case soci::data_type::dt_integer: {
+          auto value = row.get<int>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetInt(value);
+        }
+        break;
+        case soci::data_type::dt_long_long: {
+          auto value = row.get<long long>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetInt64(value);
+        }
+        break;
+        case soci::data_type::dt_unsigned_long_long: {
+          auto value = row.get<unsigned long long>(i);
+          if (pMaxCollector_) {
+            pMaxCollector_->updateMaxValue(columnName, value);
+          }
+          valueVal.SetUint64(value);
+        }
+        break;
+        case soci::data_type::dt_date: {
+          // It looks like design bug in soci - for `dt_date`, it returns std::tm, which doesn't have mmilliseconds, but DB 'datetime' has milliseconds.
+          // Don't know if it is possible to get a string representation for 'dt_date' type.
+          // The problem for maxCollector, if milliseconds value is not stored as a maximum, then when running query with 'select ... datetimeColumn > maxValue', 
+          // it will be always at least one record since DB has milliseconds "maxValue.milliseconds".
+          // For a workaround in the string representation for 'dt_date', add '999' for maxCollector (won't work for cases where time precision is important).
+          std::tm when = row.get<std::tm>(i);
+
+          char strWhen[128];
+          std::strftime(strWhen, sizeof(strWhen), "%Y-%m-%d %H:%M:%S", &when);
```

Comment:
I think the return value should be checked here. In case it's 0, the data in array (strWhen) is undefined.

Reply:
Fixed.